### PR TITLE
fix: sanitize HTML before dangerouslySetInnerHTML

### PR DESCRIPTION
**Pull request by yashwant938**
Harden client-side rendering paths that use dangerouslySetInnerHTML by sanitizing HTML with a small allowlist sanitizer.\n\n- Add utils/sanitizeHtml helper (removes scripts, event handlers, unsafe URLs; enforces rel=noopener noreferrer for target=_blank).\n- Apply sanitization to release notes, upgrade carousel details, tern tooltip docs, manual upgrades, and CRUD info modal.\n- Add unit tests for sanitizer behavior.\n\nThis reduces DOM-XSS risk if upstream HTML or interpolated content is compromised.

Fixes #`Issue Number`  
_or_  
Fixes `Issue URL`
> [!WARNING]  
> _If no issue exists, please create an issue first, and check with the maintainers if the issue is valid._

## Automation

/ok-to-test tags=""

### :mag: Cypress test results

> [!CAUTION]  
> If you modify the content in this section, you are likely to disrupt the CI result for your PR.

## Communication
Should the DevRel and Marketing teams inform users about this change?
- [ ] Yes
- [ ] No

## Summary by CodeRabbit

* **New Features**
  * Added a client-side HTML sanitizer to ensure unsafe HTML is stripped before rendering.

* **Bug Fixes**
  * Sanitized dynamic HTML across product descriptions, upgrade prompts, release notes and tooltips to prevent unsafe rendering.

* **Style**
  * Improved rendering of inline code blocks and adjusted URL presentation in upgrade messages for clearer display.

* **Tests**
  * Added unit tests validating HTML sanitization behavior and safety guarantees.

### CODE REVIEW

File: app/client/src/components/BottomBar/ManualUpgrades.tsx [hunk 1]
```diff
@@ -156,20 +157,17 @@ function ManualUpgrades(props: {
...
+            `/app/${applicationSlug}/${pageSlug}-${pageId}`,
+          );
+
+          return sanitizeHtml(createMessage(formatter.bind(null, cleanUrl)));
+        }),
         disclaimer: {
           severity: "MODERATE",
-          desc: createMessage(CLEAN_URL_UPDATE.disclaimer),
+          desc: sanitizeHtml(createMessage(CLEAN_URL_UPDATE.disclaimer)),
         },
```
[the remark this answers is not in the document]
Comment (yashwant938) [reply app/client/src/components/BottomBar/ManualUpgrades.tsx:174]: Good catch. I didn’t whitelist style in the sanitizer (that would weaken the XSS hardening). Instead I:

Removed the inline [style="..."](https://file+.vscode-resource.vscode-cdn.net/home/yash/.vscode/extensions/openai.chatgpt-0.4.71-linux-x64/webview/#) from the CLEAN_URL_UPDATE message template so sanitization doesn’t strip anything meaningful ([messages.ts](https://file+.vscode-resource.vscode-cdn.net/home/yash/.vscode/extensions/openai.chatgpt-0.4.71-linux-x64/webview/#) now uses <code>${url}</code>).
Moved the intended presentation (wrapping + line-height: 22px) into the existing [StyledList code { ... }](https://file+.vscode-resource.vscode-cdn.net/home/yash/.vscode/extensions/openai.chatgpt-0.4.71-linux-x64/webview/#) styles in [ManualUpgrades.tsx](https://file+.vscode-resource.vscode-cdn.net/home/yash/.vscode/extensions/openai.chatgpt-0.4.71-linux-x64/webview/#), so the <code> block still looks the same without relying on inline styles.
This is included in commit b0cb647ffb.

File: app/client/src/utils/sanitizeHtml.test.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
+import sanitizeHtml from "./sanitizeHtml";
+
+describe("sanitizeHtml", () => {
+  it("removes script tags", () => {
+    const dirty = "<p>ok<script>alert(1)</script></p>";
+    const clean = sanitizeHtml(dirty);
+    expect(clean).toContain("<p>ok</p>");
+    expect(clean).not.toContain("<script");
+    expect(clean).not.toContain("alert(1)");
+  });
```
[the remark this answers is not in the document]
Comment (yashwant938) [reply app/client/src/utils/sanitizeHtml.test.ts:10]: - Changed `sanitizeElement` in `app/client/src/utils/sanitizeHtml.ts` so dangerous tags like `<script>` are **dropped entirely (element + children)** instead of being “unwrapped” into the parent.
  - Before: `<script>alert(1)</script>` would unwrap and leave `alert(1)` as visible text.
  - After: the whole `<script>...</script>` block is removed, so `alert(1)` never appears in the output.
- Added a drop-list (`DROP_TAGS_WITH_CONTENT`) that includes `script` (and also `style/iframe/object/embed/noscript`, plus `template/svg/math/foreignobject` as defense-in-depth).

**Commit:** `55c06f7a13`

File: app/client/src/utils/sanitizeHtml.test.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    expect(clean).not.toContain("href=");
+  });
+
+  it("adds rel=noopener noreferrer for target=_blank", () => {
+    const dirty = '<a href="https://example.com" target="_blank">x</a>';
+    const clean = sanitizeHtml(dirty);
+    expect(clean).toContain('target="_blank"');
+    expect(clean).toContain('rel="noopener noreferrer"');
+  });
+});
```
[the remark this answers is not in the document]
Comment (yashwant938) [reply app/client/src/utils/sanitizeHtml.test.ts]: - Added test cases covering the reviewer’s list:
  - `<style>` removal
  - `<iframe>`, `<object>`, `<embed>` removal (and ensuring their contents don’t leak through)
  - SVG/Math vectors (dropped entirely as defense-in-depth)
  - `data:` URI handling:
    - strip `data:` from `<a href>`
    - allow safe `data:image/...` on ``
  - HTML comments removal
  - Empty and non-string inputs (`""`, whitespace, `null`, `undefined`, numbers)
  - Positive cases: allowed tags/attributes are preserved
  - `style` attribute stripping (e.g. `<code style="...">` becomes `<code>...`)
- These updates are in `app/client/src/utils/sanitizeHtml.test.ts`.

**Commit:** `55c06f7a13`

File: app/client/src/utils/sanitizeHtml.ts [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+  const template = document.createElement("template");
+  template.innerHTML = dirtyHtml;
+
+  for (const child of Array.from(template.content.childNodes)) {
+    sanitizeNode(child);
+  }
+
+  return template.innerHTML;
+}
```
[the remark this answers is not in the document]
Comment (yashwant938) [reply app/client/src/utils/sanitizeHtml.ts:218]: Hardened `sanitizeHtml` to *drop entirely* the parsing-context/namespace footgun tags (`template`, `svg`, `math`, `foreignObject`, `noscript`) along with `script/style/iframe/object/embed`, instead of trying to unwrap/sanitize them. This avoids the nested-`<template>` `.content` gap and SVG/MathML namespace confusion routes.
- **What we added:** Expanded unit tests to cover these vectors (plus comments, data URI rules, positive allowlist cases, non-string input).
- **Files:** `app/client/src/utils/sanitizeHtml.ts`, `app/client/src/utils/sanitizeHtml.test.ts`
- **Commit:** `55c06f7a13`

Pasteable resolve reply:
> Agree DOMPurify is the gold-standard for HTML sanitization and handles mXSS/namespace edge cases better than a custom implementation. For this PR I kept things dependency-free/scope-limited, but hardened the sanitizer to avoid the specific bypass vectors called out by dropping `template/svg/math/foreignObject/noscript` (and other high-risk tags) entirely rather than attempting to unwrap/sanitize them, and added tests covering these cases. If maintainers prefer, I can follow up with a DOMPurify-based implementation.

File: app/client/src/utils/sanitizeHtml.ts [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+  if (!DEFAULT_ALLOWED_TAGS.has(tag)) {
+    const parent = el.parentNode;
+    if (!parent) return;
+
+    while (el.firstChild) {
+      parent.insertBefore(el.firstChild, el);
+    }
+    parent.removeChild(el);
+    return;
+  }
```
[the remark this answers is not in the document]
Comment (yashwant938) [reply app/client/src/utils/sanitizeHtml.ts:134]: - In `app/client/src/utils/sanitizeHtml.ts`, we added an explicit drop-list (`DROP_TAGS_WITH_CONTENT`) for tags that must be removed **with their contents** (`script`, `style`, `iframe`, `object`, `embed`, `noscript`, plus `template/svg/math/foreignobject` as defense-in-depth).
- `sanitizeElement` now checks that list first and does `el.parentNode?.removeChild(el); return;` so nothing gets unwrapped into the parent.
- The original “unwrap children for disallowed tags” behavior remains only for non-dangerous unknown tags where we want to preserve text content.

**Commit:** `55c06f7a13`
